### PR TITLE
Split score multiplier and unranked label colours

### LINKED ISSUE #2313: Unranked mods should have a score multiplier of 1x and another color
Will do this. I want to know the multiplier when I have an unranked mod on. I just can't decide on a color.

### PR DESCRIPTION
**Pull request by Joehuu**
Resolves #2313. ~Chose yellow for lower multiplier, the color of the diff inc section.~
Edit: Chose `(Unranked)` label to be blue, the same colour as the special section.

### CODE REVIEW

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 1]
```diff
@@ -98,15 +100,24 @@ private void updateMods()
...
+
+            RankedLabel.Text = null;
+            if (!ranked)
+            {
+                RankedLabel.Text += " (Unranked)";
+                RankedLabel.FadeColour(RankedColour, 200);
+            }
+            else
+            {
+                RankedLabel.Text = null;
```
Comment (Aergwyn) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]: Unnecessary. `RankedLabel.Text` will be `null` at this point.

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 2]
```diff
@@ -98,15 +100,24 @@ private void updateMods()
...
                 MultiplierLabel.FadeColour(HighMultiplierColour, 200);
             else if (multiplier < 1.0)
                 MultiplierLabel.FadeColour(LowMultiplierColour, 200);
             else
                 MultiplierLabel.FadeColour(Color4.White, 200);
+
+            RankedLabel.Text = null;
+            if (!ranked)
+            {
+                RankedLabel.Text += " (Unranked)";
```
Comment (Aergwyn) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]: `=` should be enough here to set. Also the leading whitespace could be removed and the label could have an appropriate distance instead.

File: osu.Game.Tests/Visual/TestCaseMods.cs [hunk 3]
```diff
@@ -182,13 +182,13 @@ private void testMultiplierTextColour(Mod mod, Color4 colour)
             checkLabelColor(Color4.White);
         }
 
-        private void testMultiplierTextUnranked(Mod mod)
+        private void testRankedTextUnranked(Mod mod)
         {
-            AddAssert("check for ranked", () => !modSelect.MultiplierLabel.Text.EndsWith(unranked_suffix));
+            AddAssert("check for ranked", () => !modSelect.RankedLabel.Text.EndsWith(unranked_suffix));
```
Comment (Aergwyn) [inline osu.Game.Tests/Visual/TestCaseMods.cs]: I think we can check for `Equals` now that it's a dedicated label.

File: osu.Game.Tests/Visual/TestCaseMods.cs [hunk 4]
```diff
@@ -182,13 +182,13 @@ private void testMultiplierTextColour(Mod mod, Color4 colour)
             checkLabelColor(Color4.White);
         }
 
-        private void testMultiplierTextUnranked(Mod mod)
+        private void testRankedTextUnranked(Mod mod)
```
Comment (Aergwyn) [inline osu.Game.Tests/Visual/TestCaseMods.cs]: I'd even suggest naming this method `testRankedText` now.

**Comment by Aergwyn** [comment]
> Chose yellow for lower multiplier, the color of the diff inc section.

Now that you say it. I may have a weird suggestion.  
Difficulty decreasing mods are green and difficulty increasing mods are yellow. Shouldn't the multiplier colour reflect that? Right now you click the green mods and get a yellow colour and vice versa.  
Maybe it would be interesting to see how it is if you swap it so below 1 is green and above 1 is yellow.  
  
Then again I'm no UX person but the conflicting colours would be still weird imho.

**Comment by Joehuu** [comment]
That would make special mods that decrease confusing. I plan to move the mods on #2250, and to make your idea work, the special section would have to have the decreasing mods moved to the reduction section and `(Unranked)` will have to be blue. And the sections have to be renamed to `Score Decrease` and etc. Lastly I have another suggestion, shouldn’t the `Score Multiplier:` be coloured from the multiplier like on stable?

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs (hunk 2 above)
Comment (Joehuu) [reply osu.Game/Overlays/Mods/ModSelectOverlay.cs]: You said a conflicting review. If I do `RankedLabel.Text = "(Unranked)"` with no space, it overrides `RankedLabel.Text = " "` on the top. It's best that I replace only the `+=` with `=`.

Comment (Aergwyn) [reply osu.Game/Overlays/Mods/ModSelectOverlay.cs]: There is no `RankedLabel.Text = " "` though? D: Am I blind?

Comment (Joehuu) [reply osu.Game/Overlays/Mods/ModSelectOverlay.cs]: > Also the leading whitespace could be removed and the label could have an appropriate distance instead.

Maybe I'm confused. What is this asking exactly?

Comment (Aergwyn) [reply osu.Game/Overlays/Mods/ModSelectOverlay.cs]: Have `RankedLabel.Text = "(Unranked)";` and eventually move the label to the right with padding if it doesn't look appropriate.

Comment (jorolf) [reply osu.Game/Overlays/Mods/ModSelectOverlay.cs]: *margin

**Comment by Joehuu** [comment]
Nvm, your suggestion wouldn't make sense if you have a diff increase mod and reduction mod at the same time. That'll result in green(multiplier less than 1.0) with your suggestion. I plan to make `(Unranked)` blue from special though. And leave the `multiplier < 1.0` as red.

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 5]
```diff
@@ -110,14 +119,11 @@ private void updateMods()
             RankedLabel.Text = null;
```
Comment (Aergwyn) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]: Crazy idea: Have the label fixed with text/colour and just hide/show it instead of constantly changing it.

Comment (Joehuu) [reply osu.Game/Overlays/Mods/ModSelectOverlay.cs]: How do you do that? Also, having fixed colour on the label won't show the fade animation.

Comment (Aergwyn) [reply osu.Game/Overlays/Mods/ModSelectOverlay.cs]: You don't need a colour transition as it'd always be blue.
`RankedLabel.FadeIn(200)` or `RankedLabel.FadeOut(200)`, 200 being the time it takes to complete.

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 6]
```diff
@@ -351,14 +367,14 @@ public ModSelectOverlay()
                                                 Right = 20
                                             }
                                         },
-                                        new OsuSpriteText
+                                        ScoreLabel = new OsuSpriteText
                                         {
-                                            Text = @"Score Multiplier: ",
```
Comment (jorolf) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]: why did you move this to `updateMods()`?

Comment (Joehuu) [reply osu.Game/Overlays/Mods/ModSelectOverlay.cs]: Shouldn't `Score Multiplier:` change color with the multiplier like on stable?

Comment (Aergwyn) [reply osu.Game/Overlays/Mods/ModSelectOverlay.cs]: No it shouldn't (personal opinion). I think the way it looks in lazer feels cleaner.

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 7]
```diff
@@ -97,16 +100,29 @@ private void updateMods()
...
 
+            MultiplierLabel.Text = $"{multiplier:N2}x";
             if (multiplier > 1.0)
                 MultiplierLabel.FadeColour(HighMultiplierColour, 200);
             else if (multiplier < 1.0)
                 MultiplierLabel.FadeColour(LowMultiplierColour, 200);
             else
                 MultiplierLabel.FadeColour(Color4.White, 200);
+
+            RankedLabel.Text = "(Unranked)";
```
Comment (Aergwyn) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]: You can set the text at label declaration instead.

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 8]
```diff
@@ -97,16 +100,29 @@ private void updateMods()
                 ranked &= mod.Ranked;
             }
 
-            MultiplierLabel.Text = $"{multiplier:N2}x";
-            if (!ranked)
-                MultiplierLabel.Text += " (Unranked)";
+            ScoreLabel.Text = "Score Multiplier:";
```
Comment (Aergwyn) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]: This doesn't need to be set with each update. As it's static text at this point you can just set it at label declaration.

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 9]
```diff
@@ -57,6 +59,7 @@ private void load(OsuColour colours, OsuGame osu, RulesetStore rulesets, AudioMa
 
             LowMultiplierColour = colours.Red;
             HighMultiplierColour = colours.Green;
+            RankedColour = colours.Blue;
```
Comment (Aergwyn) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]: The reference is not needed. Just set the colour of the label.

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 10]
```diff
@@ -97,16 +100,29 @@ private void updateMods()
...
                 MultiplierLabel.FadeColour(HighMultiplierColour, 200);
             else if (multiplier < 1.0)
                 MultiplierLabel.FadeColour(LowMultiplierColour, 200);
             else
                 MultiplierLabel.FadeColour(Color4.White, 200);
+
+            RankedLabel.Text = "(Unranked)";
+            RankedLabel.FadeColour(RankedColour, 200);
+            RankedLabel.FadeOut(200);
+            if (!ranked)
```
Comment (Aergwyn) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]: The FadeColour shouldn't be needed so only
```csharp
if(ranked)
    RankedLabel.FadeOut(200);
else
    RankedLabel.FadeIn(200);
```
remains.

File: osu.Game.Tests/Visual/TestCaseMods.cs [hunk 11]
```diff
@@ -182,13 +182,13 @@ private void testMultiplierTextColour(Mod mod, Color4 colour)
             checkLabelColor(Color4.White);
         }
 
-        private void testMultiplierTextUnranked(Mod mod)
+        private void testRankedText(Mod mod)
         {
-            AddAssert("check for ranked", () => !modSelect.MultiplierLabel.Text.EndsWith(unranked_suffix));
+            AddAssert("check for ranked", () => !modSelect.RankedLabel.Text.Equals(unranked_suffix));
```
Comment (Joehuu) [inline osu.Game.Tests/Visual/TestCaseMods.cs]: How do you check for fade?

Comment (jorolf) [reply osu.Game.Tests/Visual/TestCaseMods.cs]: Which "fade" do you mean? Colour or Alpha? Either way you can check the property

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs (hunk 6 above)
Comment (jorolf) [reply osu.Game/Overlays/Mods/ModSelectOverlay.cs]: What does color have to do with the text of a label?

File: osu.Game.Tests/Visual/TestCaseMods.cs [hunk 12]
```diff
@@ -182,13 +182,16 @@ private void testMultiplierTextColour(Mod mod, Color4 colour)
             checkLabelColor(Color4.White);
         }
 
-        private void testMultiplierTextUnranked(Mod mod)
+        private void testRankedText(Mod mod)
         {
-            AddAssert("check for ranked", () => !modSelect.MultiplierLabel.Text.EndsWith(unranked_suffix));
+            AddWaitStep(1, "wait for fade");
+            AddAssert("check for ranked", () => modSelect.RankedLabel.Alpha.Equals(0));
```
Comment (jorolf) [inline osu.Game.Tests/Visual/TestCaseMods.cs]: Why didn't you use "==" 🤔

Comment (Frontear) [reply osu.Game.Tests/Visual/TestCaseMods.cs]: Is there a difference, besides obvious syntax?

**Review by peppy — APPROVED** [review]
